### PR TITLE
Emmc stm32 driver

### LINKED ISSUE #36375: sdmmc_stm32 disk driver will not work with eMMC
**Is your enhancement proposal related to a problem? Please describe.**
There are two distinct protocols used over the SDMMC bus: 
* JEDEC JESD84-B51A: Embedded MultiMediaCard (e.MMC), Electrical Standard (5.1A).
* SD Card "SD Specifications"
While these two standards have much in common, they also have crucial differences in the protocol (especially during device initialization). 

**Describe the solution you'd like**
The attached patch adds a new disk driver for the STM32 family of SOCs to utilize the MMC protocol instead of the SD. The 2 drivers are very similar (indeed, I converted the SDCard driver to the emmc driver with simple global edits), and it might be possible to combine them better. However, this is a simple and effective solution that will allow developers to target either type of device.

**Describe alternatives you've considered**
My initial approach was to edit the `sdmmc_stm32.c` disk driver In place. However, I understand that this could break existing users that do rely on the SD protocol.

[patch.tar.gz](https://github.com/zephyrproject-rtos/zephyr/files/6672237/patch.tar.gz)

### PR DESCRIPTION
**Pull request by nickstoughton**
Most (all?) eMMC devices use the JEDEC MMC protocol. There subtle differences
in the initialization phase between the MMC and SD protocols, and they
cannot be used interchangeably.

Signed-off-by: Nick Stoughton <nstoughton@logitech.com>

Fixes #36375

### CODE REVIEW

**Comment by carlescufi** [comment]
@nickstoughton thanks for the PR. Please format the commits as per our [commit guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines) and do not include fixup commits in the PR. Instead rebase and amend commits in order to maintain a clean set of them.

File: drivers/disk/Kconfig.sdmmc [hunk 1]
```diff
@@ -11,7 +11,12 @@ config DISK_DRIVER_SDMMC
 	help
 	  SDMMC card driver.
 
-if DISK_DRIVER_SDMMC
+config DISK_DRIVER_EMMC
+	bool "EMMC card driver"
+	help
+	  EMMC card driver.
+
+if DISK_DRIVER_SDMMC || DISK_DRIVER_EMMC
```
Comment (jfischer-no) [inline drivers/disk/Kconfig.sdmmc:19]: These changes are not necessary.

File: drivers/disk/Kconfig.sdmmc [hunk 2]
```diff
@@ -51,8 +56,17 @@ config SDMMC_STM32
 	help
 	  File system on sdmmc accessed through stm32 sdmmc.
 
+config EMMC_STM32
+	bool "STM32 EMMC driver"
```
Comment (jfischer-no) [inline drivers/disk/Kconfig.sdmmc:60]: STM32 **e**MMC driver

File: drivers/disk/Kconfig.sdmmc [hunk 3]
```diff
@@ -51,8 +56,17 @@ config SDMMC_STM32
...
+	depends on HAS_STM32CUBE
+	select USE_STM32_HAL_MMC
+	select USE_STM32_LL_SDMMC
+	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_EMMC))
+	help
+	  File system on eMMC accessed through stm32 sdmmc.
```
Comment (jfischer-no) [inline drivers/disk/Kconfig.sdmmc:66]:
```
help
  STM32 eMMC controller shim driver.
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/disk/emmc_stm32.c [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+	.name = CONFIG_SDMMC_VOLUME_NAME,
+	.ops = &stm32_emmc_ops,
+};
+
+/*
+ * Check if the card is present or not. If no card detect gpio is set, assume
+ * the card is present. If reading the gpio fails for some reason, assume the
+ * card is there.
+ */
+static bool stm32_sdmmc_card_present(struct stm32_sdmmc_priv *priv)
```
Comment (jfischer-no) [inline drivers/disk/emmc_stm32.c:227]: sorry, my comment got lost somehow. Most of the code below related to card-present and power is equal to sdmmc_stm32 code, can you please move it from sdmmc_stm32.c to a common file like mmc_stm32_common.c and then reuse in new shim driver?

Comment (nickstoughton) [reply drivers/disk/emmc_stm32.c:227]: Thinking about it, eMMC never uses card-detect, and the power gpio is better renamed reset. I can simply push a change that removes all the card detect logic, and renames `power-gpios` to `reset-gpios`, but splitting out the common code from the original SD driver runs a risk that I do not have access to any SD cards to be able to test that I did not bread the originall driver, though I agree that ultimately this is a better approach.

Please advise on how you would like to see this resolved.

Comment (erwango) [reply drivers/disk/emmc_stm32.c:227]: @nickstoughton, we'll be able to test that there is no regression on original SD driver, so you can go ahead as this makes sense.

**Comment by erwango** [comment]
@nickstoughton thanks for contributing this driver!

Do you plan to also contribute the HAL fix you mentioned in #36375 ?

Also, FYI, in general we expect drivers to be enabled on one in tree board, so we're able to at least build (and potentially test) driver in CI once it is merged. From what I checked there is no such board in tree, and the only (STM32 based) one I know with eMMC support would be stm32h745i_disco. So we'll try to get this board upstreamed when possible, unless you're able to do it on your side.

**Pull request closed without merging**